python/pyiceberg/expressions/literals.py

```diff
@@ -356,18 +326,26 @@ def _(self, type_var: DecimalType) -> Literal[Decimal]:
         return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
 
 
-class DateLiteral(Literal[int]):
-    def __init__(self, value: int):
-        super().__init__(value, int)
+class DateLiteral(Literal[date]):
+    def __init__(self, value: date):
+        super().__init__(value, date)
```

Comment:
Sorry, I'm not sure when this changed, but I don't think that we want to store a date in a `DateLiteral`. We want to normalize that to the standard representation, days from unix epoch.

Reply:
I thought we discussed this earlier to make sure that we can do:
```
def literal(value: T) -> Literal[T]
```
Otherwise you'll get:
```
def literal(value: date) -> Literal[int]
```

Reply:
Does `@overload` help with this since we can return a `Literal[int]` for a date argument?

Reply:
Yes, overload solves it for creating the types. But then we're stuck at the next step:
```
def _to_literal(value: Union[U, Literal[T]]) -> Literal[T]:
    if isinstance(value, Literal):
        return value
    else:
        return literal(value)
```
We use this to do things like `EqualTo('co', 'val')`. Actually, we want to combine the type here, but this isn't possible with Python generics. It should look at the mapping of `literal`, and figure out the result type. For certain types, such as `date` we get a `Literal[int]`. 

When using `T`:
```
def _to_literal(value: Union[T, Literal[T]]) -> Literal[T]:
    if isinstance(value, Literal):
        return value
    else:
        return literal(value)
```
We get `Literal[date]` instead.